### PR TITLE
only include entry points defined in source

### LINKED ISSUE #59696: Incorrect entries being shown after calling FindEntryPoints
**Version Used**: 17.2 Preview 2

**Steps to Reproduce**:
The project system is calling [FindentryPoints](https://github.com/dotnet/roslyn/blob/587388c52c4aefb5010e972c619222f4025182f1/src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb#L26) when populating the Startup Object property in the Project Properties UI (see [here](https://github.com/dotnet/project-system/blob/6d9a0248f63682e4d8071534329d6770f6e2770a/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/StartupObjectsEnumProvider.cs#L87)). This is done for both the C# and VB scenarios, with the distinction that the VB scenario will be called with the flag `findFormsOnly = true`. However, calling that function returns a set of unexpected entries.

**Expected Behavior**:
Only returns forms entries.

**Actual Behavior**:
image

### PR DESCRIPTION
**Pull request by jmarolf**
Fixes #59696

In winforms apps on .NET Core we are given a global namespace that contains symbols defined as metadata. This never happened on .NET Framework. We now filter out all symbols not in metadata so we don't show incorrect entry points.

### CODE REVIEW

File: src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb [hunk 1]
```diff
@@ -30,7 +31,12 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.ProjectSystemShim
         End Function
 
         Public Overrides Sub VisitNamedType(symbol As INamedTypeSymbol)
-            ' It's a form if it Inherits System.Windows.Forms.Form. 
+            ' Metadata symbols cannot be entry points.
+            If Not symbol.IsFromSource Then
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb:35]: If we only want to walk source, an easier way is to just get the source assembly namespace for a project and walk down that. Then you skip metadata symbols entirely.

Comment (jmarolf) [reply src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb]: yeah this was a lot simpler, thanks!

File: src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb [hunk 2]
```diff
@@ -25,12 +25,13 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.ProjectSystemShim
 
         Public Shared Function FindEntryPoints(symbol As INamespaceSymbol, findFormsOnly As Boolean) As IEnumerable(Of INamedTypeSymbol)
             Dim visitor = New EntryPointFinder(findFormsOnly)
-            visitor.Visit(symbol)
+            ' Only search source symbols
+            visitor.Visit(symbol.Compilation.SourceModule.GlobalNamespace)
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb:29]: We have one of these for C# too, does that also need an update?

Comment (jmarolf) [reply src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb:29]: may as well. This bug does not reproduce for C# but some future set of references may construct a global namespace that could.

**Review by JoeRobich — APPROVED** [review]

**Review by jasonmalinowski — APPROVED** [review]

**Comment by melytc** [comment]
Gentle ping on this. I see PR is set to auto-merge, but checks haven't passed completely.

**Comment by jmarolf** [comment]
@melytc integration tests have been on the floor for a week+. I can keep re-running them and hoping they eventually pass.

File: src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb [hunk 3]
```diff
@@ -25,12 +25,13 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.ProjectSystemShim
 
         Public Shared Function FindEntryPoints(symbol As INamespaceSymbol, findFormsOnly As Boolean) As IEnumerable(Of INamedTypeSymbol)
             Dim visitor = New EntryPointFinder(findFormsOnly)
-            visitor.Visit(symbol)
+            ' Only search source symbols
+            visitor.Visit(symbol.ContainingCompilation.SourceModule.GlobalNamespace)
             Return visitor.EntryPoints
         End Function
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb:31]: it saddens me that we have two different copies of this code in C#/VB :D

File: src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@ protected override bool MatchesMainMethodName(string name)
         public static IEnumerable<INamedTypeSymbol> FindEntryPoints(INamespaceSymbol symbol)
         {
             var visitor = new EntryPointFinder();
-            visitor.Visit(symbol);
+            visitor.Visit(symbol.ContainingCompilation.SourceModule.GlobalNamespace);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs:21]: doc this one as well?

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**